teps/000X-trigger-crd.md

```diff
@@ -0,0 +1,351 @@
+```
+apiVersion: v1alpha1
+kind: Trigger
+metadata:
+  name: my-repo-trigger
+spec:
+  serviceAccountName: "blah"
+  interceptors:
+    cel:
+      filter: "$(header.eventType == "push")"
+  bindings:
+    - name: key
+      value: "$(body.commit_id)"
+  template: "ref-to-my-template"
+ status:
+   address:
+      url: "el-my-svc.cluster.local" # could also be an IP address?
+   conditions:
+      status: True
+      message: "Bound to EventListener my-el"
+```
+
+#### Trigger Status
+Trigger status will report url of EventListener bound and EventListener information to 
+which it is bound.
+
+### EventListener CRD
+EventListeners expose an addressable "Sink" to which incoming events in the form of HTTP requests
+are directed and process those requests with configuration from bound Triggers. Boundnamespace acts 
+as selector based on namespace. All triggers in namespaces specified are selected can be served by
+EventListener.
+
+```
+apiVersion: v1alpha1
+kind: EventListener
+metadata:
+  name: my-repo-trigger
+spec:
+  serviceAccountName: "blah"
+  boundnamespace:
+      - foo-ns
+      - bar-ns
+```
+
+### Path based EventListener
+Inside EventListener, we will deduce Trigger to execute based on the path of the request URL.
+To refer to trigger resource foo in namespace bar, we would url ```/bar/foo```. Instead of 
+EventListener iterating through every Triggers that can be serve by EventListener, EventListener
+will directly process the Trigger refer by URL.
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this TEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Trigger CRD Implmentation
+Trigger CRD will be implemented. Trigger CRD spec will contain serviceaccount, triggerbinding, 
+triggertemplate and interceptor. ServiceAccount will be optional. If not defined, ServiceAccount
+of EventListener will be used.
+
+#### EventListener Selector Implementation
+Selector based EventListener implementation will be introduced. All namespace matching the 
+boundnamespaces in EventListener will be served by EventListener.
+
+#### Path Base EventListener Implementation
+Path base EventListener will be implemented.
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+<!--
+What are the risks of this proposal and how do we mitigate. Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the WGs or subproject.
+-->
```

Comment:
there are probably some risks here worth fleshing out, e.g.:

- what if one namespace has more requests than other namespaces and effects the performance of others
- the cluster wide event listener needs to have permission to create instances in all namespaces
- others?

Reply:
Added those.